Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,381 @@
+                textNodeList.append(currentNode);
+            if (auto newNode = nextInComposedTreeIgnoringUserAgentShadow(*currentNode))
+                currentNode = newNode;
+            else
+                break;
+        }
+
+        if (auto range = findRangeFromNodeList(query, searchRange, textNodeList, wordStartBounded, wordEndBounded))
+            return range;
 
-    return resultRange.collapsed() ? std::nullopt : std::make_optional(resultRange);
+        if (!currentNode)
+            break;
+        
+        ASSERT_WITH_MESSAGE(is_gt(treeOrder(*currentNode, searchRange.start.container)), "Scroll To Text Fragment: Current node must follow start.");
+        
+        if (!is_gt(treeOrder(*currentNode, searchRange.start.container)))
+            return std::nullopt;
+        
+        searchRange.start = { *currentNode, 0 };
+    }
+    return std::nullopt;
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#next-non-whitespace-position
+static void advanceRangeStartToNextNonWhitespace(SimpleRange& range)
+{
+    while (!range.collapsed()) {
+        auto& node = range.startContainer();
+        auto offset = range.startOffset();
+        
+        // This check is not in the spec.
+        // I believe there is an error in the spec which I have filed an issue for
+        // https://github.com/WICG/scroll-to-text-fragment/issues/189
+        if (offset == node.length()) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node)) {
+                range.start = { *newStart, 0 };
+                continue;
+            }
+            return;
+        }
+        
+        if (isNonSearchableSubtree(node)) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node))
+                range.start = { *newStart, 0 };
+            else
+                return;
+            continue;
+        }
+
+        if (!isVisibleTextNode(node)) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node))
+                range.start = { *newStart, 0 };
+            else
+                return;
+            continue;
+        }
+
+        auto string = node.textContent();
+
+        if (string.substringSharingImpl(offset, 6) == "&nbsp;"_s)
+            offset += 6;
+
+        if (string.substringSharingImpl(offset, 5) == "&nbsp"_s)
+            offset += 5;
+        
+        if (!isSpaceOrNewline(string[offset]))
+            return;
+        offset++;
+        
+        if (offset >= node.length()) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node))
+                range.start = { *newStart, 0 };
+            else
+                return;
+        } else
+            range.start = { node, offset };
+    }
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#find-a-range-from-a-text-directive
+std::optional<SimpleRange> rangeForFragment(const ParsedTextDirective parsedTextDirective, Document& document)
```

Comment:
We should call this function `findRangeFromTextDirective` like the spec says.
"fragment" has a very specific meaning in this spec and the rest of web platform
so I think it's more confusing to call this function that.